text/0000-symbol-name-mangling-v2.md

```diff
@@ -49,7 +49,7 @@ A symbol mangling scheme has a few goals, one of them essential, the rest of the
 
  - A mangled symbol should be *decodable* to some degree. That is, it is desirable to be able to tell which exact concrete instance of e.g. a polymorphic function a given symbol identifies. This is true for external tools, backtraces, or just people only having the binary representation of some piece of code available to them. With the current scheme, this kind of information gets lost in the magical hash-suffix.
 
- - It should be possible to predict the symbol name for a given source-level construct. For example, given the definition `fn foo<T>() { ... }`, the scheme should allow to construct, by hand, the symbol names for e.g. `foo<u32>` or `foo<extern fn(i32, &mut SomeStruct<(char, &str)>, ...) -> !>()`. Since the current scheme generates its hash from the values of various compiler internal data structures, not even an alternative compiler implementation could predicate the symbol name, even for simple cases.
+ - It should be possible to predict the symbol name for a given source-level construct. For example, given the definition `fn foo<T>() { ... }`, the scheme should allow to construct, by hand, the symbol names for e.g. `foo<u32>` or `foo<extern fn(i32, &mut SomeStruct<(char, &str)>, ...) -> !>()`. Since the current scheme generates its hash from the values of various compiler internal data structures, not even an alternative compiler implementation could predict the symbol name, even for simple cases.
```

Comment:
We should avoid (accidentally) providing interop guarantees. A compiler should only be guaranteed to interop between artifacts *it generated*.

Unless what we're saying is that *for a specific compiler version*, an alternative compiler should be able to generate the same symbol names, but certain details might remain implementation-specific?

Still, it's likely of little use, given that the necessary compiler-specific metadata blob won't be compatible.

Reply:
Maybe we should keep the hash but make it optional, and have the demangler hide it?
Or at least enough bits of it (1-2 base62 digits) to make it unpredictable.

Reply:
Coming back to this, 2 new ideas:

1. on the first build of a crate, generate some entropy, store it in crate and incremental metadata (so it can be reused until that crate is cleared)

2. we don't *really* need to add an extra character or two from the hash if we make the crate hashes *explicitly* and *intentionally* unpredictable, uninteropable and generally implementation-defined

In general, we could use 1. to be less "absolutely" deterministic than we otherwise would be (e.g. `TypeId`), and avoid accidental interop between two *different* builds where `Cargo` happens to use the exact same `rustc` flags (`-Cmetadata` specifically) with the same source.

Reply:
> if we make the crate hashes explicitly and intentionally unpredictable

Would this make the build nondeterministic, in the sense used by [reproducible-builds.org](https://reproducible-builds.org/)

Reply:
@zackw Oh, I had forgotten about builds like that, I was only thinking of incremental.
In that case we can only include entropy specific to the `rustc` compiler executable (and that build would *also* have to be determinstic - can probably use the version, including the git hash).

In any case, I *think* it's enough to make sure there's no predictable way to go from `-Cmetadata` values to the crate hashes in the mangling, even if it has to be deterministic.

Reply:
As long as this safeguard is:
- encoded in a single component somewhere,
- can be omitted in user-facing demanglings,
- does not impede reproducible builds, and
- can be removed from the grammar, should we ever want to define a stable ABI, 

then I think we might be able do it?

Reply:
IMO this bulletpoint in the RFC should be rephrased and/or made an explicit anti-goal.

To summarize the comments above, I think it might be sufficient for the crate hashes to be unpredictable (*even if deterministic* for a given Rust implementation), if you only know the `rustc` command-line arguments passed by `Cargo`.

There's, however, a small risk of some tool extracting enough of these hashes to reuse them. For that, we should *explicitly* say that these are implementation-defined, change between versions, generally have no guarantees, and should *not* be used for interop.

If we *do* want to add some extra information, I suppose we can just not guarantee anything about the crate hashes and extend some of them with additional digits, or maybe generalize the scheme used to add hashes to crates and merge it with disambiguators (e.g. for `impl`)?

We can say disambiguators are implementation-defined in general, can't we?